File: langhuan-service/src/main/java/com/langhuan/controller/UserController.java
package com.langhuan.controller;

import com.langhuan.common.Result;
import com.langhuan.model.domain.TUser;
import com.langhuan.model.dto.UserLoginDTO;
import com.langhuan.service.TUserService;
import com.langhuan.utils.other.JwtUtil;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.authorization.AuthorizationDeniedException;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.logout.SecurityContextLogoutHandler;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.stream.Collectors;

@RestController
@RequestMapping(path = "/user")
public class UserController {
    private final JwtUtil jwtUtil;

    private final TUserService TUserService;

    public UserController(JwtUtil jwtUtil, TUserService TUserService) {
        this.jwtUtil = jwtUtil;
        this.TUserService = TUserService;
    }

    @PreAuthorize("hasAuthority('/user/register')")
    @PostMapping("/register")
    public Result register(@RequestBody TUser user) {
        return Result.success(TUserService.register(user));
    }

    @PreAuthorize("hasAuthority('/user/change')")
    @PostMapping("/change")
    public Result change(@RequestBody TUser user) throws AuthorizationDeniedException {
        return Result.success(TUserService.change(user));
    }

    @PostMapping("/login")
    public Result login(@RequestBody @Validated UserLoginDTO userLoginDTO, HttpServletResponse response) {
        return Result.success(TUserService.login(userLoginDTO, response));
    }

    // @PreAuthorize("hasAuthority('/user/info/view')")
    @PostMapping("/getUserInfoByToken")
    public Result getUserInfoByToken(
    ) {
        return Result.success(TUserService.getUserInfoByToken());
    }

    @PreAuthorize("hasAuthority('/user/list')")
    @PostMapping("/getUserPageList")
    public Result getUserPageList(
            @RequestParam(name = "name", required = false, defaultValue = "") String name,
            @RequestParam(name = "username", required = false, defaultValue = "") String username,
            @RequestParam(name = "gender", required = false) Integer gender,
            @RequestParam(name = "enabled", required = false) Integer enabled,
            @RequestParam(name = "pageNum", required = false, defaultValue = "1") int pageNum,
            @RequestParam(name = "pageSize", required = false, defaultValue = "10") int pageSize
    ) {

        return Result.success(TUserService.getUserPageList(name, username, gender, enabled, pageNum, pageSize));
    }

    @PreAuthorize("hasAuthority('/user/delete')")
    @PostMapping("/delete")
    public Result delete(@RequestParam(name = "id", required = true) Integer id) throws AuthorizationDeniedException {
        if (id == 1) {
            return Result.error("管理员用户无法删除");
        }
        Boolean delete = TUserService.delete(id);
        if (!delete) {
            return Result.error("删除失败");
        }
        return Result.success("删除成功");
    }

    @PreAuthorize("hasAuthority('/user/roles/view')")
    @PostMapping("/getUserRoles")
    public Result getUserRoles(@RequestParam(name = "id", required = false) Integer id) {
        return Result.success(TUserService.getUserRoles(id));
    }

    @PreAuthorize("hasAuthority('/user/roles/edit')")
    @PostMapping("/relevancyRoles")
    public Result relevancyRoles(
            @RequestParam(name = "id", required = true) Integer id,
            @RequestParam(name = "roleIds", required = true) String roleIds
    ) {
        try {
            if (roleIds.isEmpty()) {
                TUserService.relevancyRoles(id, new ArrayList<>());
            } else {
                String[] strings = roleIds.split(",");
                TUserService.relevancyRoles(id, Arrays.stream(strings).map(Integer::parseInt).collect(Collectors.toList()));
            }
        } catch (Exception e) {
            return Result.error("角色id格式错误");
        }
        return Result.success("操作成功");
    }


    //@PreAuthorize("hasAuthority('/user/list')")
    //@PreAuthorize("hasAnyRole('admin', 'normal')")
    @GetMapping("/logout")
    public Result logout(HttpServletRequest request, HttpServletResponse response) {
        // 退出登录
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        // 清除认证
        new SecurityContextLogoutHandler().logout(request, response, auth);
        return Result.success("操作成功");
    }
}

File: langhuan-server/src/main/kotlin/com/langhuan/controller/UserController.kt
package com.langhuan.controller

import com.langhuan.common.ApiLog
import com.langhuan.common.Result
import com.langhuan.model.domain.TUser
import com.langhuan.model.dto.UserLoginDTO
import com.langhuan.service.TUserService
import com.langhuan.utils.other.JwtUtil
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.security.authorization.AuthorizationDeniedException
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.logout.SecurityContextLogoutHandler
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(path = ["/user"])
class UserController(
    private val jwtUtil: JwtUtil,
    private val tUserService: TUserService
) {

    companion object {
        private val log = LoggerFactory.getLogger(UserController::class.java)
    }

    // @PreAuthorize("hasAuthority('/user/register')")
    @PostMapping("/register")
    fun register(@RequestBody user: TUser): Result<Any> {
        return Result.success(tUserService.register(user))
    }

    @PreAuthorize("hasAuthority('/user/change')")
    @PostMapping("/change")
    @Throws(AuthorizationDeniedException::class)
    fun change(@RequestBody user: TUser): Result<Any> {
        return Result.success(tUserService.change(user))
    }

    @PostMapping("/login")
    fun login(@RequestBody @Validated userLoginDTO: UserLoginDTO, response: HttpServletResponse): Result<Any> {
        return Result.success(tUserService.login(userLoginDTO, response))
    }

    // @PreAuthorize("hasAuthority('/user/info/view')")
    @PostMapping("/getUserInfoByToken")
    fun getUserInfoByToken(): Result<Any> {
        return Result.success(tUserService.getUserInfoByToken())
    }

    @PreAuthorize("hasAuthority('/user/list')")
    @PostMapping("/getUserPageList")
    fun getUserPageList(
        @RequestParam(name = "name", required = false, defaultValue = "") name: String,
        @RequestParam(name = "username", required = false, defaultValue = "") username: String,
        @RequestParam(name = "gender", required = false) gender: Int?,
        @RequestParam(name = "enabled", required = false) enabled: Int?,
        @RequestParam(name = "pageNum", required = true, defaultValue = "1") pageNum: Int,
        @RequestParam(name = "pageSize", required = true, defaultValue = "10") pageSize: Int
    ): Result<Any> {
        return Result.success(tUserService.getUserPageList(name, username, gender, enabled, pageNum, pageSize))
    }

    @ApiLog(apiName = "删除用户", description = "根据用户ID删除用户信息")
    @PreAuthorize("hasAuthority('/user/delete')")
    @PostMapping("/delete")
    @Throws(AuthorizationDeniedException::class)
    fun delete(@RequestParam(name = "id", required = true) id: Int): Result<Any> {
        if (id == 1) {
            return Result.error("管理员用户无法删除")
        }
        val delete = tUserService.delete(id)
        if (!delete) {
            return Result.error("删除失败")
        }
        return Result.success("删除成功")
    }

    @PreAuthorize("hasAuthority('/user/roles/view')")
    @PostMapping("/getUserRoles")
    fun getUserRoles(@RequestParam(name = "id", required = false) id: Int?): Result<Any> {
        return Result.success(tUserService.getUserRoles(id))
    }

    @PreAuthorize("hasAuthority('/user/roles/edit')")
    @PostMapping("/relevancyRoles")
    fun relevancyRoles(
        @RequestParam(name = "id", required = true) id: Int,
        @RequestParam(name = "roleIds", required = true) roleIds: String
    ): Result<Any> {
        try {
            if (roleIds.isEmpty()) {
                tUserService.relevancyRoles(id, ArrayList())
            } else {
                val strings = roleIds.split(",").toTypedArray()
                tUserService.relevancyRoles(
                    id,
                    strings.map { it.toInt() }
                )
            }
        } catch (e: Exception) {
            return Result.error("角色id格式错误")
        }
        return Result.success("操作成功")
    }

    // @PreAuthorize("hasAuthority('/user/list')")
    // @PreAuthorize("hasAnyRole('admin', 'normal')")
    @PostMapping("/logout")
    fun logout(request: HttpServletRequest, response: HttpServletResponse): Result<Any> {
        // 退出登录
        val auth = SecurityContextHolder.getContext().authentication
        // 清除认证
        SecurityContextLogoutHandler().logout(request, response, auth)
        return Result.success("操作成功")
    }

    // @PreAuthorize("hasAuthority('/user/list')")
    @PostMapping("/getUserRoleListById")
    fun getUserRoleListById(@RequestParam(name = "roleId", required = false) roleId: Int?): Result<Any> {
        return Result.success(tUserService.getUserRoleListById(roleId as Int))
    }
}
